zproject/backends.py

```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+
+        if not request_state:
+            logging.info("Sign in with Apple failed: missing state parameter.")
+            raise AuthMissingParameter(self, 'state')
+
+        formatted_request_state = "apple_auth_" + request_state
+        redis_data = get_dict_from_redis(redis_client, "apple_auth_{token}",
+                                         formatted_request_state)
+        if redis_data is None:
+            logging.info("Sign in with Apple failed: bad state token.")
+            raise AuthStateForbidden(self)
+
+        for param, value in redis_data.items():
+            if param in self.standard_relay_params:
+                self.strategy.session_set(param, value)
+        return request_state
+
+    def is_native_flow(self) -> bool:
+        return self.strategy.request_data().get('native_flow', False)
+
+    def decode_id_token(self, id_token: str) -> Dict[str, Any]:
+        '''
+        Decode and validate JWT token from apple and return payload including user data.
+        We override this, because upstream implementation does not support the use of "Bundle ID"
+        which should be used as audience to decode id_token in the iOS native auth flow.
+        We also take advantage of this override to improve error handling and exceptions.
+        '''
+        if self.is_native_flow():
+            audience = self.setting("BUNDLE_ID")
+        else:
+            audience = self.setting("SERVICES_ID")
+
+        try:
+            kid = jwt.get_unverified_header(id_token).get('kid')
+            public_key = RSAAlgorithm.from_jwk(self.get_apple_jwk(kid))
+            decoded = jwt.decode(id_token, key=public_key,
+                                 audience=audience, algorithm="RS256")
+        except PyJWTError:
+            raise AuthFailed(self, "Token validation failed")
+
+        return decoded
+
+    def auth_complete(self, *args: Any, **kwargs: Any) -> Optional[HttpResponse]:
+        request_data = self.strategy.request_data()
+        if self.is_native_flow():
+            if 'id_token' not in request_data:
+                raise JsonableError(_("Missing id_token parameter"))
+
+            for param in self.standard_relay_params:
+                self.strategy.session_set(param, request_data.get(param))
+
+            # We should get the subdomain from the hostname of the request.
+            self.strategy.session_set('subdomain', get_subdomain(self.strategy.request))
+
+            try:
+                result = self.do_auth(request_data['id_token'], *args, **kwargs)
+                return result
+            except (AuthFailed, AuthCanceled) as e:
+                # AuthFailed is a general "failure" exception from PSA that we should convert
+                # to None return value here to avoid getting tracebacks.
+                # AuthCanceled is raised in the Apple backend implementation in PSA
+                # in certain cases, though AuthFailed would have been more correct.
```

Comment:
Why is AuthFailed more correct than AuthCanceled?  The comment may not need to change, but I'd like to have a clear understanding of why this comment is accurate, since it seems like an important detail.

Reply:
From upstream:
```
class AuthCanceled(AuthException):
    """Auth process was canceled by user."""
```
```
class AuthFailed(AuthException):
    """Auth process failed for some reason."""
```